server/game/costs/PlayCardResourceCost.ts

```diff
@@ -22,9 +22,11 @@ export class PlayCardResourceCost<TContext extends AbilityContext = AbilityConte
             return false;
         }
 
+        const costAdjusterFromAbility = 'costAdjuster' in context.ability ? context.ability.costAdjuster : null;
```

Comment:
This is a bit brittle, if we change the name of this property on PlayCardAction then it will silently break. If we do `Contract.assertTrue(context.ability.isCardPlayed())` then type narrowing should make `context.ability.costAdjuster` accessible

Reply:
Doesn't work: `Property 'costAdjuster' does not exist on type 'PlayerOrCardAbility'`

Reply:
Ahh right because PlayerOrCardAbility is still a JS file. I actually just tried something out and made a discovery - I thought that it wasn't possible to use type predicates like this correctly with a JS class but it turns out you can. If you can make the following tweak to the code the above will work.

In PlayerOrCardAbility change `isCardPlayed()` to be this (including the jsdoc):

```typescript
// TODO: refactor the other methods to also be type predicates
/**
 * Indicates whether a card is played as part of the resolution this ability
 * @returns {this is PlayCardAction}
 */
isCardPlayed() {
    return false;
}
```

then make a similar change in PlayCardAction:

```typescript
public override isCardPlayed(): this is PlayCardAction {
    return true;
}
```

Then it should work!

Reply:
That does let us use `context.ability.costAdjuster` directly, but unfortunately importing `PlayCardAction` into PlayerOrCardAbility.js so we can declare the type predicate causes (what I think is) an import error:
`TypeError: Class extends value [object Object] is not a constructor or null
    at Object.<anonymous> (/home/ennis/MyStuff/misc/programming/Github/forceteki/server/game/core/ability/PlayerAction.js:6:28)`

Reply:
Figured it out. I took a hint from isAttackAction() further down and did an import right in the return type declaration.